File: 0155-min-stack/0155-min-stack.cpp
class MinStack {
public:
    MinStack() {
        
    }
    stack<long long> st;
    long long mini = INT_MAX;
    void push(int value) {
        if(st.empty()){
            st.push(value);
            mini = value;
        }
        else{
            if(mini <= value){
                st.push(value);
            }
            else{
                long long newVal = 2ll * value - mini;
                st.push(newVal);
                mini = value;
            }
        }
    }
    
    void pop() {
        long long x = st.top();
        st.pop();
        if(x < mini){
            mini = 2 * mini - x;
        }
    }
    
    int top() {
        if(st.top() < mini){
            return mini;
        }
        else{
            return st.top();
        }
    }
    
    int getMin() {
        return mini;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(value);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 0155-min-stack/0155-min-stack.py
class MinStack(object):

    def __init__(self):
        self.st = []
        self.mini = 0
    def push(self, value):
        if not self.st:
            self.st.append(value)
            self.mini = value
        else:
            if value >= self.mini:
                self.st.append(value)
            else:
                newVal = 2*value - self.mini
                self.st.append(newVal)
                self.mini = value
    def pop(self):
        self.x = self.st[-1]
        if self.mini <= self.x:
            self.st.pop()
        else:
            self.mini = 2*self.mini - self.x
            self.st.pop()

    def top(self):
        self.x = self.st[-1]
        if self.mini <= self.x:
            return self.x
        else:
            return self.mini
        
        

    def getMin(self):
        return self.mini
        """
        :rtype: int
        """
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(value)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()